### PR TITLE
[RFC] Adding overrides for max cache seq length

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

#1364 

#### Changelog

This PR:
- Adds support for overriding the maximum sequence length used when setting up KV-caches. 
- Adds support for correctly setting up caches for self-attention, cross-attention, and fusion layers by exposing `encoder` and `decoder` `max_seq_len` args. These arguments are exposed to the top-level transformer class (i.e. `TransformerDecoder,` `DeepFusionModel`, so that the API for all models remains the same `model.setup_caches(bsz, dtype, encoder_max_seq_len, decoder_max_seq_len)`
- We also remove the use of `input_pos` to update and retrieve from the KV-cache. Instead, the KV cache tracks its own position.

image

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [x] I have added an example to docs or docstrings;

### CODE REVIEW

**Comment by SalmanMohammadi** [comment]
What's up with the eleuther eval tests? They're passing locally for me.

File: torchtune/modules/model_fusion/_fusion.py
Comment (SalmanMohammadi) [inline torchtune/modules/model_fusion/_fusion.py:323]: had to do this due to circular imports, open to something more sane looking

File: torchtune/modules/attention.py
Comment (SalmanMohammadi) [inline torchtune/modules/attention.py:283]: What would be a sensible way to only use the cache for generation? For PPO, I need to run two forward passes with the same model, all under `no_grad`
1) During generation, with cache enabled
2) During trajectory estimation, using the same model with LoRA adapters disabled, but **without KV cacheing**

Then, we obviously don't want to use KV cacheing when `requires_grad=True` during the loss calculations.
My current solution is to wrap all generation code into `torch.inference_mode()`. Then, we check here for `if x.is_inference()` to only use the cache during inference, and not in any other case. 

Currently this is broken for compile (see issue here https://github.com/pytorch/pytorch/issues/135439#issuecomment-2337454492), but hopefully a fix by 2.5. Like most other places in the repo, could we version check for compile + generation?

@pbontrager @joecummings

Comment (SalmanMohammadi) [reply torchtune/modules/attention.py:283]: On reflection this isn't important for this PR and we can address in a follow up when it is needed for PPO.

File: torchtune/modules/model_fusion/_fusion.py
Comment (ebsmothers) [inline torchtune/modules/model_fusion/_fusion.py:6]: Is this also cause of circular imports?

File: torchtune/modules/model_fusion/_fusion.py
Comment (ebsmothers) [reply torchtune/modules/model_fusion/_fusion.py:323]: I'm confused, what's the circular import here? Seems to me like TransformerDecoder should not be depending on fusion, right?

File: torchtune/modules/transformer.py
Comment (ebsmothers) [inline torchtune/modules/transformer.py:660]: So it might be too much of a sledgehammer to do it this way, but in my mind it's also more explicit about the condition we actually care about.
```suggestion
        has_encoder_layers = any(
            isinstance(m, TransformerCrossAttentionLayer) for m in self.modules()
        )
```

File: torchtune/modules/transformer.py
Comment (ebsmothers) [inline torchtune/modules/transformer.py:787]: At this point I would offload all of this (plus the seq len validation just above) into `self._validate_inputs` or something. Either way should do this check prior to calling token embeddings though

File: torchtune/utils/_generation.py
Comment (ebsmothers) [inline torchtune/utils/_generation.py:130]: bsz > 1?

File: torchtune/utils/_generation.py
Comment (ebsmothers) [inline torchtune/utils/_generation.py:163]: 😕

**Review by ebsmothers — COMMENTED** [review]
Overall it looks reasonable to me, will leave it to @pbontrager or @joecummings for the final sign-off

**Comment by SalmanMohammadi** [comment]
Thanks so much for the reviews @ebsmothers. Will address once #1424 lands and things are merged.

**Review by joecummings — APPROVED** [review]
No concerns, but pls address @ebsmothers comments.

**Pull request merged**
